Source/diablo.cpp

```diff
@@ -2202,4 +1963,281 @@ void diablo_color_cyc_logic()
 	}
 }
 
+void helpKeyPressed()
+{
+	if (helpflag) {
+		helpflag = false;
+	} else if (stextflag != STORE_NONE) {
+		ClearPanel();
+		AddPanelString(_("No help available"), true); /// BUGFIX: message isn't displayed
+		AddPanelString(_("while in stores"), true);
+		track_repeat_walk(false);
+	} else {
+		invflag = false;
+		chrflag = false;
+		sbookflag = false;
+		spselflag = false;
+		if (qtextflag && leveltype == DTYPE_TOWN) {
+			qtextflag = false;
+			stream_stop();
+		}
+		questlog = false;
+		automapflag = false;
+		msgdelay = 0;
+		gamemenu_off();
+		DisplayHelp();
+		doom_close();
+	}
+}
+
+void itemInfoKeyPressed()
+{
+	if (pcursitem != -1) {
+		sprintf(
+		    tempstr,
+		    "IDX = %i  :  Seed = %i  :  CF = %i",
+		    items[pcursitem].IDidx,
+		    items[pcursitem]._iSeed,
+		    items[pcursitem]._iCreateInfo);
+		NetSendCmdString(1 << myplr, tempstr);
+	}
+	sprintf(tempstr, "Numitems : %i", numitems);
+	NetSendCmdString(1 << myplr, tempstr);
+}
+
+void inventoryKeyPressed()
+{
+	if (stextflag != STORE_NONE)
+		return;
+	invflag = !invflag;
+	if (!chrflag && !questlog && PANELS_COVER) {
+		if (!invflag) { // We closed the invetory
+			if (MouseX < 480 && MouseY < PANEL_TOP) {
+				SetCursorPos(MouseX + 160, MouseY);
+			}
+		} else if (!sbookflag) { // We opened the invetory
+			if (MouseX > 160 && MouseY < PANEL_TOP) {
+				SetCursorPos(MouseX - 160, MouseY);
+			}
+		}
+	}
+	sbookflag = false;
+}
+
+void characterSheetKeyPressed()
+{
+	if (stextflag != STORE_NONE)
+		return;
+	chrflag = !chrflag;
+	if (!invflag && !sbookflag && PANELS_COVER) {
+		if (!chrflag) { // We closed the character sheet
+			if (MouseX > 160 && MouseY < PANEL_TOP) {
+				SetCursorPos(MouseX - 160, MouseY);
+			}
+		} else if (!questlog) { // We opened the character sheet
+			if (MouseX < 480 && MouseY < PANEL_TOP) {
+				SetCursorPos(MouseX + 160, MouseY);
+			}
+		}
+	}
+	questlog = false;
+}
+
+void questLogKeyPressed()
+{
+	if (stextflag != STORE_NONE)
+		return;
+	if (!questlog) {
+		StartQuestlog();
+	} else {
+		questlog = false;
+	}
+	if (!invflag && !sbookflag && PANELS_COVER) {
+		if (!questlog) { // We closed the quest log
+			if (MouseX > 160 && MouseY < PANEL_TOP) {
+				SetCursorPos(MouseX - 160, MouseY);
+			}
+		} else if (!chrflag) { // We opened the character quest log
+			if (MouseX < 480 && MouseY < PANEL_TOP) {
+				SetCursorPos(MouseX + 160, MouseY);
+			}
+		}
+	}
+	chrflag = false;
+}
+
+void displaySpellsKeyPressed()
+{
+	if (stextflag != STORE_NONE)
+		return;
+	chrflag = false;
+	questlog = false;
+	invflag = false;
+	sbookflag = false;
+	if (!spselflag) {
+		DoSpeedBook();
+	} else {
+		spselflag = false;
+	}
+	track_repeat_walk(false);
+}
+
+void spellBookKeyPressed()

```

Comment:
This and other seem to duplicate the code from `GameAction` handling use for gamepad support, e.g. see `GameActionType_TOGGLE_SPELL_BOOK`.

Reply:
I just moved the code from one place to the other in the same file. If there is a duplication it was done by `GameAction`, not this PR. It doesn't seem trivial though:
```
	if (stextflag != STORE_NONE)
		return;
	sbookflag = !sbookflag;
	if (!chrflag && !questlog && PANELS_COVER) {
		if (!sbookflag) { // We closed the invetory
			if (MouseX < 480 && MouseY < PANEL_TOP) {
				SetCursorPos(MouseX + 160, MouseY);
			}
		} else if (!invflag) { // We opened the invetory
			if (MouseX > 160 && MouseY < PANEL_TOP) {
				SetCursorPos(MouseX - 160, MouseY);
			}
		}
	}
	invflag = false;
```
vs
```
			if (BlurInventory()) {
				invflag = false;
				spselflag = false;
				sbookflag = !sbookflag;
			}
```
In any case, I think that could go in another PR.

Reply:
Fair enough

Reply:
I think that this whole panel management code needs a refactor though.